release: 11.0.0 (#3906)
rome/tools: CHANGELOG.md

# Rome changelog

## 11.0.0

### CLI

#### BREAKING CHANGES

- the argument `--no-colors` has been removed, in favor of `--color=off`

#### Other changes

- `init` command now adds the `$schema` property to the generated `rome.json` file,
  if` rome` is installed inside the `node_modules` folder. Check the [guide] to know how
  manually update this property to an existing project.
- Added `--semicolons` argument to the `rome format`, where users can specify whether
  they want semicolons or not in their source code.
- Rome now exists with an error code if it doesn't process any file.
- Fixed how the maximum number of diagnostics is calculated [#3869](https://github.com/rome/tools/pull/3869).

> - Rome now prints the total number of files that caused errors.

Reviewer: ```suggestion
- Rome now prints the total number of errors caused in the files.
```

Reviewer 2: Apologies for butting in. I just noticed there’s a typo. I think “exists” should be “exists” in the above snippet.

Author: @rbargholz code reviews are always welcome :)

- Rome now correctly traverses symbolic links, it correctly detects possible loops and
  doesn't stop the traversal;
- the new global argument `--colors` now accepts `force`, useful to force colors when
  Rome is spawned as a process;

### Configuration

- Added a `$schema` property, where users can add the JSON schema. This feature provides
  auto-completion and descriptions of all fields of the configuration file.
- Added a new `files.ignore` option where users can ignore files across the board.
-

### Editors

- The extension is published [Open VSX](https://open-vsx.org/) too.
- The extension now correctly resolves the version installed via the `node_modules` folder
- Fixed an issue where diagnostics were not updated after a change to the configuration file (#3724)[https://github.com/rome/tools/pull/3724]

### Formatter

#### BREAKING CHANGES

- Fixed incompatibility issues with Prettier [#3531](https://github.com/rome/tools/issues/3531)
  - [#3686](https://github.com/rome/tools/pull/3686)
  - [#3732](https://github.com/rome/tools/pull/3732)

>   - [#3824](https://github.com/rome/tools/pull/3824)

Reviewer 3: I believe it's typo in pull request number

- Fixed an issue where infinite parentheses were wrongly inserted [#3735](https://github.com/rome/tools/issues/3735)
- Better formatting for `jestEach` templates

#### Other changes

- Added support for omitting semicolons.

### Linter

- Fixed false positives emitted by `noUselessFragments` [#3668](https://github.com/rome/tools/issues/3668)
- Fixed `noArrayIndexKey` where some cases were not detected [#3670](https://github.com/rome/tools/issues/3670)
- Fixed false positives emitted by `noConstAssign` [#3728](https://github.com/rome/tools/issues/3728)
- Fixed false positives emitted by `noShoutyConstants` [#3867](https://github.com/rome/tools/issues/3867)
- Fixed false positives emitted by `noUnusedVariables` [#3779](https://github.com/rome/tools/issues/3779)
- Fixed `noUndeclaredVariables` where some cases were not detected [#3798](https://github.com/rome/tools/issues/3798)

> #### New rules
> 
> - [noAccessKey](https://docs.rome.tools/lint/rules/noAccessKey)
> - [noBannedTypes](https://docs.rome.tools/lint/rules/noBannedTypes)
> - [noConditionalAssignment](https://docs.rome.tools/lint/rules/noConditionalAssignment)
> - [noConstAssign](https://docs.rome.tools/lint/rules/noConstAssign)
> - [noConstEnum](https://docs.rome.tools/lint/rules/noConstEnum)
> - [noConstructorReturn](https://docs.rome.tools/lint/rules/noConstructorReturn)
> - [noDistractingElements](https://docs.rome.tools/lint/rules/noDistractingElements)
> - [noDupeKeys](https://docs.rome.tools/lint/rules/noDupeKeys)
> - [noEmptyInterface](https://docs.rome.tools/lint/rules/noEmptyInterface)
> - [noExplicitAny](https://docs.rome.tools/lint/rules/noExplicitAny)
> - [noExtraNonNullAssertion](https://docs.rome.tools/lint/rules/noExtraNonNullAssertion)
> - [noHeaderScope](https://docs.rome.tools/lint/rules/noHeaderScope)
> - [noInvalidConstructorSuper](https://docs.rome.tools/lint/rules/noInvalidConstructorSuper)
> - [noNonNullAssertion](https://docs.rome.tools/lint/rules/noNonNullAssertion)
> - [noPrecisionLoss](https://docs.rome.tools/lint/rules/noPrecisionLoss)
> - [noSetterReturn](https://docs.rome.tools/lint/rules/noSetterReturn)
> - [noStringCaseMismatch](https://docs.rome.tools/lint/rules/noStringCaseMismatch)
> - [noUnsafeFinally](https://docs.rome.tools/lint/rules/noUnsafeFinally)
> - [noVar](https://docs.rome.tools/lint/rules/noVar)
> - [noVoidTypeReturn](https://docs.rome.tools/lint/rules/noVoidTypeReturn)
> - [useConst](https://docs.rome.tools/lint/rules/useConst)
> - [useDefaultSwitchClauseLast](https://docs.rome.tools/lint/rules/useDefaultSwitchClauseLast)
> - [useFlatMap](https://docs.rome.tools/lint/rules/useFlatMap)
> - [useNumericLiterals](https://docs.rome.tools/lint/rules/useNumericLiterals)
> - [useAriaPropsForRole](https://docs.rome.tools/lint/rules/useAriaPropsForRole)
> - [useAriaPropTypes](https://docs.rome.tools/lint/rules/useAriaPropTypes)

Reviewer 4: I rewrote this section to add more details on the changes of #3908:
````suggestion
#### Rules

The following rules have been stabilized:
- `nusery/useFlatMap` -> `complexity/useFlatMap`
- `nusery/useValidForDirection` -> `correctness/useValidForDirection`
- `nusery/noExplicitAny` -> `suspicious/noExplicitAny`
- `nusery/noConstAssign` -> `correctness/noConstAssign`

These rules are all recommended, so they will be enabled by default. If you had enabled them manually from the `nursery` group, you can simply remove those entries from your configuration file.

The following rules have been renamed:
- `correctness/noMultipleSpacesInRegularExpressionLiterals` -> `complexity/noMultipleSpacesInRegularExpressionLiterals`
- `style/useOptionalChain` -> `complexity/useOptionalChain`
- `correctness/noUselessFragments` -> `complexity/noUselessFragments`
- `correctness/noDelete` -> `performance/noDelete`
- `correctness/useSingleCaseStatement` -> `style/useSingleCaseStatement`
- `correctness/useWhile` -> `style/useWhile`
- `correctness/noArguments` -> `style/noArguments`
- `correctness/noAsyncPromiseExecutor` -> `suspicious/noAsyncPromiseExecutor`
- `correctness/noCommentText` -> `suspicious/noCommentText`
- `correctness/noCompareNegZero` -> `suspicious/noCompareNegZero`
- `correctness/noDebugger` -> `suspicious/noDebugger`
- `correctness/noDoubleEquals` -> `suspicious/noDoubleEquals`
- `correctness/noShadowRestrictedNames` -> `suspicious/noShadowRestrictedNames`
- `correctness/noSparseArray` -> `suspicious/noSparseArray`
- `correctness/noUnsafeNegation` -> `suspicious/noUnsafeNegation`
- `correctness/useValidTypeof` -> `suspicious/useValidTypeof`
- `correctness/noArrayIndexKey` -> `suspicious/noArrayIndexKey`
- `correctness/noCatchAssign` -> `suspicious/noCatchAssign`
- `correctness/noDupeArgs` -> `suspicious/noDuplicateParameters`
- `correctness/noFunctionAssign` -> `suspicious/noFunctionAssign`
- `correctness/noImportAssign` -> `suspicious/noImportAssign`
- `correctness/noLabelVar` -> `suspicious/noLabelVar`
- `correctness/noRestrictedGlobals` -> `nursery/noRestrictedGlobals`
- `nursery/noDupeKeys` -> `nursery/noDuplicateObjectKeys`

If you were not changing the severity level of any of these rules in your configuration file, or suppressing a diagnostic emitted by those rules using suppression comments, you do not have to do anything. But if you did, Rome will now emit diagnostics for the parts of your configuration or suppression comments you need to update.

The following rules are no longer recommended:
- `style/noImplicitBoolean`
- `style/noNegationElse`
- `style/useBlockStatements`
- `style/useShorthandArrayType`
- `correctness/useSingleCaseStatement` / `style/useSingleCaseStatement`
- `style/noShoutyConstants`

The styling decisions imposed by these rules were not deemed to be idiomatic enough in the JavaScript ecosystem to be enabled by default. If you do want to enforce those rules in your project, you will have to enable them manually in you configuration file:

```json
{
  "linter": {
    "rules": {
        "style": {
            "useBlockStatements": "warn"
        }
    }
  }
}
```

Finally, the following new rules have been introduced to the nursery group in this release:
- [`nursery/noAccessKey`](https://docs.rome.tools/lint/rules/noAccessKey)
- [`nursery/noConditionalAssignment`](https://docs.rome.tools/lint/rules/noConditionalAssignment)
- [`nursery/noConstEnum`](https://docs.rome.tools/lint/rules/noConstEnum)
- [`nursery/noConstructorReturn`](https://docs.rome.tools/lint/rules/noConstructorReturn)
- [`nursery/noDistractingElements`](https://docs.rome.tools/lint/rules/noDistractingElements)
- [`nursery/noDuplicateObjectKeys`](https://docs.rome.tools/lint/rules/noDuplicateObjectKeys)
- [`nursery/noEmptyInterface`](https://docs.rome.tools/lint/rules/noEmptyInterface)
- [`nursery/noExtraNonNullAssertion`](https://docs.rome.tools/lint/rules/noExtraNonNullAssertion)
- [`nursery/noHeaderScope`](https://docs.rome.tools/lint/rules/noHeaderScope)
- [`nursery/noNonNullAssertion`](https://docs.rome.tools/lint/rules/noNonNullAssertion)
- [`nursery/noPrecisionLoss`](https://docs.rome.tools/lint/rules/noPrecisionLoss)
- [`nursery/noRedundantUseStrict`](https://docs.rome.tools/lint/rules/noRedundantUseStrict)
- [`nursery/noSetterReturn`](https://docs.rome.tools/lint/rules/noSetterReturn)
- [`nursery/noStringCaseMismatch`](https://docs.rome.tools/lint/rules/noStringCaseMismatch)
- [`nursery/noUnsafeFinally`](https://docs.rome.tools/lint/rules/noUnsafeFinally)
- [`nursery/noVoidTypeReturn`](https://docs.rome.tools/lint/rules/noVoidTypeReturn)
- [`nursery/useDefaultSwitchClauseLast`](https://docs.rome.tools/lint/rules/useDefaultSwitchClauseLast)
- [`nursery/useNumericLiterals`](https://docs.rome.tools/lint/rules/useNumericLiterals)
- [`nursery/useAriaPropTypes`](https://docs.rome.tools/lint/rules/useAriaPropTypes)
- [`nursery/useAriaPropsForRole`](https://docs.rome.tools/lint/rules/useAriaPropsForRole)
- [`nursery/noVar`](https://docs.rome.tools/lint/rules/noVar)
- [`nursery/useConst`](https://docs.rome.tools/lint/rules/useConst)

Feel free to try them out by manually enabling them in you configuration to help us debug potential issues with these new rules.
````

- [`nursery/noHeaderScope`](https://docs.rome.tools/lint/rules/noHeaderScope)
- [`nursery/noNonNullAssertion`](https://docs.rome.tools/lint/rules/noNonNullAssertion)
- [`nursery/noPrecisionLoss`](https://docs.rome.tools/lint/rules/noPrecisionLoss)
- [`nursery/noRedundantUseStrict`](https://docs.rome.tools/lint/rules/noRedundantUseStrict)
- [`nursery/noSetterReturn`](https://docs.rome.tools/lint/rules/noSetterReturn)
- [`nursery/noStringCaseMismatch`](https://docs.rome.tools/lint/rules/noStringCaseMismatch)
- [`nursery/noUnsafeFinally`](https://docs.rome.tools/lint/rules/noUnsafeFinally)
- [`nursery/noVoidTypeReturn`](https://docs.rome.tools/lint/rules/noVoidTypeReturn)
- [`nursery/useDefaultSwitchClauseLast`](https://docs.rome.tools/lint/rules/useDefaultSwitchClauseLast)
- [`nursery/useNumericLiterals`](https://docs.rome.tools/lint/rules/useNumericLiterals)
- [`nursery/useAriaPropTypes`](https://docs.rome.tools/lint/rules/useAriaPropTypes)
- [`nursery/useAriaPropsForRole`](https://docs.rome.tools/lint/rules/useAriaPropsForRole)
- [`nursery/noVar`](https://docs.rome.tools/lint/rules/noVar)
- [`nursery/useConst`](https://docs.rome.tools/lint/rules/useConst)

> Feel free to try them out by manually enabling them in you configuration to help us debug potential issues with these new rules.

Reviewer 5: ```suggestion
Please give them a try by manually enabling them in your configuration and please share your feedback on the rule, diagnostics, and code fixes.
```

- Fixed an issue where infinite parentheses were wrongly inserted [#3735](https://github.com/rome/tools/issues/3735)
- Better formatting for `jestEach` templates

#### Other changes

- Added [support](https://docs.rome.tools/configuration/#javascriptformattersemicolon) for omitting semicolons.

### Linter

- Fixed false positives emitted by `noUselessFragments` [#3668](https://github.com/rome/tools/issues/3668)
- Fixed `noArrayIndexKey` where some cases were not detected [#3670](https://github.com/rome/tools/issues/3670)
- Fixed false positives emitted by `noConstAssign` [#3728](https://github.com/rome/tools/issues/3728)
- Fixed false positives emitted by `noShoutyConstants` [#3867](https://github.com/rome/tools/issues/3867)
- Fixed false positives emitted by `noUnusedVariables` [#3779](https://github.com/rome/tools/issues/3779)
- Fixed `noUndeclaredVariables` where some cases were not detected [#3798](https://github.com/rome/tools/issues/3798)

#### Rules

The following rules have been stabilized:
- `nusery/useFlatMap` -> `complexity/useFlatMap`
- `nusery/useValidForDirection` -> `correctness/useValidForDirection`
- `nusery/noExplicitAny` -> `suspicious/noExplicitAny`

> - `nusery/noConstAssign` -> `correctness/noConstAssign`

Reviewer 6: `nusery` -> `nursery`

- `nusery/noConstAssign` -> `correctness/noConstAssign`

These rules are all recommended, so they will be enabled by default. If you had enabled them manually from the `nursery` group, you can simply remove those entries from your configuration file.

> The following rules have been renamed:
> - `correctness/noMultipleSpacesInRegularExpressionLiterals` -> `complexity/noMultipleSpacesInRegularExpressionLiterals`

Reviewer 4: Sorry I forgot to the last rename in the list:
```suggestion
The following rules have been renamed:
- `a11y/useBlankTarget` -> `a11y/noBlankTarget`
- `correctness/noMultipleSpacesInRegularExpressionLiterals` -> `complexity/noMultipleSpacesInRegularExpressionLiterals`
```

# Rome changelog

## 11.0.0

### CLI

#### BREAKING CHANGES

- the argument `--no-colors` has been removed, in favor of `--color=off`

#### Other changes

> - `init` command now adds the `$schema` property to the generated `rome.json` file,
>   if` rome` is installed inside the `node_modules` folder. Check the [guide] to know how
>   manually update this property to an existing project.

Reviewer 5: ```suggestion
- The `init` command now adds the `$schema` property to the generated `rome.json` file
  if` rome` is installed inside the `node_modules` folder. Follow [this guide] to add the `$schema` property
  manually in a project with an existing `rome.json` file.
```

> - Added `--semicolons` argument to the `rome format`, where users can specify whether
>   they want semicolons or not in their source code.

Reviewer 5: We use the term `option` in our documentation.  
```suggestion
- A new `--semicolons` option that configures if the formatter prints semicolons at the end of every statement (default) or at the beginning of statements when necessary to prevent ASI failures.
```

> - Rome now exists with an error code if it doesn't process any file.

Reviewer 5: ```suggestion
- Rome now exits with an error code if it doesn't process any file.
```

- Fixed how the maximum number of diagnostics is calculated [#3869](https://github.com/rome/tools/pull/3869).
- Rome now prints the total number of files that caused errors.

> - Rome now correctly traverses symbolic links, it correctly detects possible loops and
>   doesn't stop the traversal;

Reviewer 5: ```suggestion
- Rome now traverses symbolic links and emits warnings if it detects loops.
```

What does "doesn't stop the traversal" mean?

> - the new global argument `--colors` now accepts `force`, useful to force colors when

Reviewer 5: ```suggestion
- You can force color output using the new global `--colors` option with the value `force`. Forcing color output can be useful if you spawn Rome as a subprocess. 
```

Why are some items terminated with a semicolon and others with a dot?

Author: > Why are some items terminated with a semicolon and others with a dot?

It's my academic brain that puts `;` at the end of each bullet point 😆 we can call it muscle memory!

  Rome is spawned as a process;

### Configuration

> - Added a `$schema` property, where users can add the JSON schema. This feature provides

Reviewer 5: ```suggestion
- Added the JSON schema `$schema` property. The schema enables auto-completion by editors and...
```

  auto-completion and descriptions of all fields of the configuration file.

> - Added a new `files.ignore` option where users can ignore files across the board.

Reviewer 5: ```suggestion
- Added a new `files.ignore` option where users can ignore files across tools.
```

> -

Reviewer 5: ```suggestion

```

### Editors

> - The extension is published [Open VSX](https://open-vsx.org/) too.

Reviewer 5: ```suggestion
- We also publish Rome to [Open VSX](https://open-vsx.org/).
```

> - The extension now correctly resolves the version installed via the `node_modules` folder

Reviewer 5: ```suggestion
- The extension now resolves the Rome version installed in the `node_modules` folder.
```

- Fixed an issue where diagnostics were not updated after a change to the configuration file (#3724)[https://github.com/rome/tools/pull/3724]

### Formatter

#### BREAKING CHANGES

- Fixed incompatibility issues with Prettier [#3531](https://github.com/rome/tools/issues/3531)
  - [#3686](https://github.com/rome/tools/pull/3686)
  - [#3732](https://github.com/rome/tools/pull/3732)
  - [#3824](https://github.com/rome/tools/pull/3824)
- Fixed an issue where infinite parentheses were wrongly inserted [#3735](https://github.com/rome/tools/issues/3735)
- Better formatting for `jestEach` templates

#### Other changes

> - Added support for omitting semicolons.

Reviewer 5: Link to a PR or documentation page. It's difficult to understand what this is about.
